cloud/previews/previews.go

```diff
@@ -0,0 +1,178 @@
+// nolint:revive
+func AppendStackPreviewLogs(client *cloud.Client, timeout time.Duration, opts AppendStackPreviewLogsOpts) (map[string]string, error) {
+	return nil, errors.E("not implemented")
+}
+
+// CreateStackPreview creates a new stack preview in the cloud
+// nolint:revive
+func CreateStackPreview(client *cloud.Client, timeout time.Duration, opts CreateStackPreviewOpts) (map[string]string, error) {
+	return nil, errors.E("not implemented")
+}
+
+// CreatePreview creates a new preview in the cloud
+func CreatePreview(client *cloud.Client, timeout time.Duration, opts CreatePreviewOpts) (*CreatedPreview, error) {
+	if err := opts.validate(); err != nil {
+		return nil, err
+	}
+
+	payload := cloud.CreatePreviewPayloadRequest{
+		UpdatedAt:       opts.UpdatedAt,
+		Technology:      opts.Technology,
+		TechnologyLayer: opts.TechnologyLayer,
+		ReviewRequest:   opts.ReviewRequest,
+		Metadata:        opts.Metadata,
+		Stacks:          []cloud.PreviewStack{},
+	}
+
+	previewStacksMap := map[string]RunContext{}
+	for _, run := range opts.Runs {
+		previewStacksMap[run.Stack.ID] = run
+	}
+
+	// loop over all affected stacks, if an item is present in the
+	// previewStacksMap, use the preview status and cmd from there
+	for _, affectedStack := range opts.AffectedStacks {
+		tags := []string{}
+		if len(affectedStack.Tags) > 0 {
+			tags = affectedStack.Tags
+		}
+
+		stack := cloud.PreviewStack{
+			PreviewStatus:   PreviewStatusAffected,
+			Repository:      opts.Repository,
+			Path:            affectedStack.Dir.String(),
+			MetaID:          strings.ToLower(affectedStack.ID),
+			MetaName:        affectedStack.Name,
+			MetaDescription: affectedStack.Description,
+			MetaTags:        tags,
+			DefaultBranch:   opts.DefaultBranch,
+			Cmd:             []string{},
+		}
+
+		if previewStack, found := previewStacksMap[affectedStack.ID]; found {
+			stack.PreviewStatus = PreviewStatusPending
+			stack.Cmd = previewStack.Cmd
+		}
+		payload.Stacks = append(payload.Stacks, stack)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	res, err := client.CreatePreview(ctx, opts.OrgUUID, payload)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(res.Stacks) != len(opts.AffectedStacks) {
+		return nil, errors.E("the backend respond with an invalid number of stacks in the deployment, got %d, expected %d",
+			len(res.Stacks), len(opts.AffectedStacks),
+			err,
+		)
+	}
+
+	stacks := map[string]string{}
+	for _, r := range res.Stacks {
+		if r.MetaID == "" {
+			return nil, errors.E("backend returned empty meta_id")
+		}
+		stacks[r.MetaID] = r.StackPreviewID
+	}
+
+	return &CreatedPreview{
+		ID:                    res.PreviewID,
+		StackPreviewsByMetaID: stacks,
+	}, nil
+}
+
+func (o CreatePreviewOpts) validate() error {
+	errs := errors.L()
+
+	if len(o.AffectedStacks) == 0 || len(o.Runs) == 0 {
+		errs.Append(errors.E("no affected stacks or runs provided"))
+	}
```

Comment:
A deployment is created even when no stacks are selected. What was decided in the case of previews? is this never going to happen? is this going to be handled by the GHA script?

Reply:
Good question, I will think about this one and get back to you.

Reply:
I think the preview should be created even if there are no selected stacks, so I removed the validation.